src/perfrecord.cpp

```diff
@@ -71,10 +74,47 @@ static QStringList sudoOptions(const QString& sudoBinary)
     return options;
 }
 
+static bool privsAlreadyElevated()
+{
+    std::ifstream ifs;
+    int i;
+    ifs.open("/proc/sys/kernel/kptr_restrict");
+    bool isElevated = false;
+    if (ifs) {
+        ifs >> i;
+        isElevated = i == 0;
+    }
+    if (!isElevated) {
+        return false;
+    }
+
+    ifs.close();
+    ifs.clear();
+    ifs.open("/proc/sys/kernel/perf_event_paranoid");
+    if (ifs) {
+        ifs >> i;
+        isElevated &= i == -1;
+    }
+    if (!isElevated) {
+        return false;
+    }
+
+    const mode_t required = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH; // 755
+    struct stat buf;
+    auto checkPerms = [required, &buf](const char* path) {
+        return stat(path, &buf) == 0 //
+            && ((buf.st_mode & 07777) & required) == required;
+    };
+    static const auto paths = {"/sys/kernel/debug", "/sys/kernel/debug/tracing"};
+    isElevated &= std::all_of(paths.begin(), paths.end(), checkPerms);
+
+    return isElevated;
+}
+
 void PerfRecord::startRecording(bool elevatePrivileges, const QStringList& perfOptions, const QString& outputPath,
                                 const QStringList& recordOptions, const QString& workingDirectory)
 {
-    if (elevatePrivileges && geteuid() != 0) {
+    if (elevatePrivileges && !privsAlreadyElevated() && geteuid() != 0) {
```

Comment:
check the geteuid first as it's cheaper

Reply:
Isn't it less probable that one would be running a gui as root?